### PR TITLE
Val_loss every N steps, model loading from pre-trained, hyperpmts reporting. 

### PR DESCRIPTION
**Pull request by krish240574**
In train_diffusion_prior.py
- Validation loss is getting every N steps, for a small batch of 100 embeddings
- Model loading implemented
- hyperparameters logging to wandb for each run
- Some code cleanup in the report_cosine_sims() function. 

I can't figure out, why the addition I made to README.md shows up again here.

### CODE REVIEW

File: train_diffusion_prior.py [hunk 1]
```diff
@@ -125,6 +127,13 @@ def train(image_embed_dim,
     # Create save_path if it doesn't exist
     if not os.path.exists(save_path):
         os.makedirs(save_path)
+        
+    # Load pre-trained model from DPRIOR_PATH
+    if RESUME:
```
Comment (rom1504) [inline train_diffusion_prior.py:132]: That's assuming the default model parameters are used

It would be better to :
* Save the hp parameters (all the kwargs of diffusionprior) into:
  * Wandb
  * The checkpoint (as another dict besides state dict)

That way we would know the params both from wandb and from the checkpoint

**Comment by rom1504** [comment]
I advise in general you use another branch that main to avoid merging issue (make one new branch for each new feature)

The metric change lgtm

For the hp param saving more is needed, left an inline comment

File: train_diffusion_prior.py (hunk 1 above)
Comment (krish240574) [reply train_diffusion_prior.py:132]: ok, gotcha. Let me fix that.

Comment (krish240574) [reply train_diffusion_prior.py:133]: Fixed. Please take a look.

File: train_diffusion_prior.py [hunk 2]
```diff
@@ -252,20 +253,31 @@ def main():
...
+    config = ({"learning_rate": args.learning_rate,
       "architecture": args.wandb_arch,
       "dataset": args.wandb_dataset,
       "Learning Rate":args.learning_rate,
       "Weight Decay":args.weight_decay,
       "Max Gradient Clipping Norm":args.max_grad_norm,
       "Batch size":args.batch_size,
-      "epochs": args.num_epochs
-      })
+      "epochs": args.num_epochs,
```
Comment (rom1504) [inline train_diffusion_prior.py:264]: can you use these params in the resume part ?

Comment (krish240574) [reply train_diffusion_prior.py:274]: Ok, fixed - the hyperparameters of the pretrained model are now being used - only in case of RESUME.

**Comment by nousr** [comment]
can we store the current position of the embedding reader when saving the model. That way it's possible to resume training and be able to pick up where you left off so to speak.

File: train_diffusion_prior.py (hunk 2 above)
Comment (rom1504) [reply train_diffusion_prior.py:274]: No you didn't understand what I mean.

In resume you are not using the hyperparameters from the checkpoint
You are still using the parameters from argparse
https://github.com/lucidrains/DALLE2-pytorch/blob/6ef5872a35ff01c55da80c6cf9b3c4ee5ffb31c0/train_diffusion_prior.py#L107 there

When doing that, you will notice it would be much more convenient to have a subdict for each network with exactly the same names as the constructor of the network so let's do that as well

Comment (krish240574) [reply train_diffusion_prior.py:309]: Fixed, do take a look .

File: train_diffusion_prior.py [hunk 3]
```diff
@@ -237,26 +298,44 @@ def main():
...
-    wandb.init(
-      entity=args.wandb_entity,
-      project=args.wandb_project,
-      config={
-      "learning_rate": args.learning_rate,
+    
+    config = ({"learning_rate": args.learning_rate,
       "architecture": args.wandb_arch,
       "dataset": args.wandb_dataset,
+      "Learning Rate":args.learning_rate,
```
Comment (rom1504) [inline train_diffusion_prior.py:305]: duplicated

File: train_diffusion_prior.py [hunk 4]
```diff
@@ -237,26 +298,44 @@ def main():
...
+      "Weight Decay":args.weight_decay,
+      "Max Gradient Clipping Norm":args.max_grad_norm,
+      "Batch size":args.batch_size,
       "epochs": args.num_epochs,
+      "Diffusion_Prior_Network":{"dpn_depth":args.dpn_depth,
+      "dpn_dim_head":args.dpn_dim_head,
+      "dpn_heads":args.dpn_heads},
+      "Diffusion_Prior":{"dp_condition_on_text_encodings": args.dp_condition_on_text_encodings,
```
Comment (rom1504) [inline train_diffusion_prior.py:313]: let's use the same names as the constructor

File: train_diffusion_prior.py [hunk 5]
```diff
@@ -41,16 +41,64 @@ def eval_model(model,device,image_reader,text_reader,start,end,batch_size,loss_t
         avg_loss = (total_loss / total_samples)
         wandb.log({f'{phase} {loss_type}': avg_loss})
 
-def save_model(save_path, state_dict):
+def save_diffusion_model(save_path, state_dict):
     # Saving State Dict
     print("====================================== Saving checkpoint ======================================")
     torch.save(state_dict, save_path+'/'+str(time.time())+'_saved_model.pth')
+    
+def load_diffusion_model(dprior_path,
```
Comment (rom1504) [inline train_diffusion_prior.py:49]: let's put that in dalle2_pytorch.py so it can be reused

File: train_diffusion_prior.py [hunk 6]
```diff
@@ -156,9 +214,9 @@ def train(image_embed_dim,
             if(int(time.time()-t) >= 60*save_interval):
                 t = time.time()
 
-                save_model(
+                save_diffusion_model(
                     save_path,
-                    dict(model=diffusion_prior.state_dict(), optimizer=optimizer.state_dict(), scaler=scaler.state_dict()))
+                    dict(model=diffusion_prior.state_dict(), optimizer=optimizer.state_dict(), scaler=scaler.state_dict(),hyperpmts = config))
```
Comment (rom1504) [inline train_diffusion_prior.py:219]: let's put  that in the function

**Comment by rom1504** [comment]
@nousr do you feel like finishing up this PR?
krish will be a bit busy in the next days

File: train_diffusion_prior.py (hunk 5 above)
Comment (krish240574) [reply train_diffusion_prior.py]: Fixed.

File: train_diffusion_prior.py (hunk 6 above)
Comment (krish240574) [reply train_diffusion_prior.py:171]: Fixed.

File: train_diffusion_prior.py (hunk 3 above)
Comment (krish240574) [reply train_diffusion_prior.py:257]: Fixed.

File: dalle2_pytorch/dalle2_pytorch.py [hunk 7]
```diff
@@ -72,6 +72,50 @@ def pad_tuple_to_length(t, length, fillvalue = None):
         return t
     return (*t, *((fillvalue,) * remain_length))
 
+# Diffusion prior model loading and saving helper functions
+
+def load_diffusion_model(dprior_path, image_embed_dim, device, wandb_entity, wandb_project):
```
Comment (rom1504) [inline dalle2_pytorch/dalle2_pytorch.py:77]: can you remove the wandb params and move image_embed_dim param in the checkpoint ?

File: train_diffusion_prior.py (hunk 6 above)
Comment (rom1504) [reply train_diffusion_prior.py:171]: looks like you changed the function, but the call still need to be adapted

File: train_diffusion_prior.py (hunk 2 above)
Comment (krish240574) [reply train_diffusion_prior.py:260]: Quick point - having the same names as the constructor might not work - to do a **dpn_config , because the DPN constructor uses a parameter from the dp_config(i.e  dp_normformer).

File: train_diffusion_prior.py (hunk 6 above)
Comment (krish240574) [reply train_diffusion_prior.py:171]: aargh, let me fix that

Comment (krish240574) [reply train_diffusion_prior.py]: Done.

File: dalle2_pytorch/dalle2_pytorch.py (hunk 7 above)
Comment (krish240574) [reply dalle2_pytorch/dalle2_pytorch.py:77]: Ok, will do .

File: train_diffusion_prior.py (hunk 4 above)
Comment (krish240574) [reply train_diffusion_prior.py]: fixed
